consensus/hotstuff/eventhandler/event_handler_v2_test.go

```diff
@@ -0,0 +1,527 @@
+
+	initView    uint64
+	endView     uint64
+	votingBlock *model.Block
+	qc          *flow.QuorumCertificate
+	newview     *model.NewViewEvent
+}
+
+func (es *EventHandlerV2Suite) SetupTest() {
+	finalized, curView := uint64(3), uint64(6)
+
+	es.paceMaker = initPaceMaker(es.T(), curView)
+	es.forks = NewForks(es.T(), finalized)
+	es.persist = &mocks.Persister{}
+	es.persist.On("PutStarted", mock.Anything).Return(nil)
+	es.blockProducer = &BlockProducer{}
+	es.communicator = &mocks.Communicator{}
+	es.communicator.On("BroadcastProposalWithDelay", mock.Anything, mock.Anything).Return(nil)
+	es.communicator.On("SendVote", mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return(nil)
+	es.committee = NewCommittee()
+	es.voteAggregator = &mocks.VoteAggregatorV2{}
+	es.voter = NewVoter(es.T(), finalized)
+	es.validator = NewBlacklistValidator(es.T())
+	es.notifier = &notifications.NoopConsumer{}
+
+	eventhandler, err := eventhandler.NewEventHandlerV2(
+		zerolog.New(os.Stderr),
+		es.paceMaker,
+		es.blockProducer,
+		es.forks,
+		es.persist,
+		es.communicator,
+		es.committee,
+		es.voteAggregator,
+		es.voter,
+		es.validator,
+		es.notifier)
+	require.NoError(es.T(), err)
+
+	es.eventhandler = eventhandler
+
+	es.initView = curView
+	es.endView = curView
+	// voting block is a block for the current view, which will trigger view change
+	es.votingBlock = createBlockWithQC(es.paceMaker.CurView(), es.paceMaker.CurView()-1)
+	es.qc = &flow.QuorumCertificate{
+		BlockID:   es.votingBlock.BlockID,
+		View:      es.votingBlock.View,
+		SignerIDs: nil,
+		SigData:   nil,
+	}
+	es.newview = &model.NewViewEvent{
+		View: es.votingBlock.View + 1, // the vote for the voting blocks will trigger a view change to the next view
+	}
+}
+
+func (es *EventHandlerV2Suite) markInvalidProposal(blockID flow.Identifier) {
+	es.validator.invalidProposals[blockID] = struct{}{}
+}
+
+// a QC for current view triggered view change
+func (es *EventHandlerV2Suite) TestQCBuiltViewChanged() {
+	// voting block exists
+	es.forks.blocks[es.votingBlock.BlockID] = es.votingBlock
+
+	// a qc is built
+	qc := createQC(es.votingBlock)
+
+	// new qc is added to forks
+	// view changed
+	// I'm not the next leader
+	// haven't received block for next view
+	// goes to the new view
+	es.endView++
+	// not the leader of the newview
+	// don't have block for the newview
+	// over
+
+	err := es.eventhandler.OnQCConstructed(qc)
+	require.NoError(es.T(), err, "if a vote can trigger a QC to be built,"+
+		"and the QC triggered a view change, then start new view")
+	require.Equal(es.T(), es.endView, es.paceMaker.CurView(), "incorrect view change")
+}
+
+// a QC for future view triggered view change
+func (es *EventHandlerV2Suite) TestQCBuiltFutureViewChanged() {
+	// voting block exists
+	curView := es.paceMaker.CurView()
+
+	b1 := createBlockWithQC(curView, curView-1)
+	b2 := createBlockWithQC(curView+1, curView)
+	b3 := createBlockWithQC(curView+2, curView+1)
```

Comment:
This test is great. Lets make sure we test with a correct scenario. The case you tested with cannot occur in reality, because having seen block b2 would already advance the view. Same for b3. 

The key point where the view does not advance is when the `qc.view` is not increasing. This would be a "fast-forward" attack by two byzantine nodes, that don't wait for the proposer of b1 to publish its block but just skip him. It is important that other consensus nodes don't just blindly follow and instead vote and the leader waits and attempts to construct a qc. 
```suggestion
// TestQCBuiltFutureViewChanged tests the handling of a "fast-forward" attack by two byzantine nodes: 
// The byzantine nodes don't wait for the previous proposer to publish its block but just skip him. 
// It is important that other consensus nodes don't just blindly follow and instead vote and the leader waits and attempts to construct a qc. 
func (es *EventHandlerV2Suite) TestQCBuiltFutureViewChanged() {
	// voting block exists
	curView := es.paceMaker.CurView()

	b1 := createBlockWithQC(curView, curView-1)
	b2 := createBlockWithQC(curView+1, curView-1)
	b3 := createBlockWithQC(curView+2, curView-1)
```